enterprise/jmx-monitoring/src/main/java/io/crate/beans/NodeInfo.java

```diff
@@ -18,35 +18,110 @@
     }
 
     @Override
     public long getClusterStateVersion() {
-        return clusterStateVersion.getAsLong();
+        return clusterState.get().version();
+    }
+
+    public ShardStats getShardStats() {
+        var total = 0;
+        var replicas = 0;
+        var unassigned = 0;
+        var primaries = 0;
+        var cs = this.clusterState.get();
+        var localNodeId = cs.nodes().getLocalNodeId();
+        var isMasterNode = cs.nodes().isLocalNodeElectedMaster();
+        var routingTable = cs.getRoutingTable();
+        if (routingTable != null) {
+            for (var shardRouting : routingTable.allShards()) {
+                if (localNodeId.equals(shardRouting.currentNodeId())) {
+                    total++;
+                    if (shardRouting.primary()) {
+                        primaries++;
+                    } else {
+                        replicas++;
+                    }
+                } else if (isMasterNode && shardRouting.unassigned()) {
+                    unassigned++;
+                }
+            }
+        }
+        return new ShardStats(total, primaries, replicas, unassigned);
+    }
+
+    @Override
+    public List<ShardInfo> getShardInfo() {
+        var cs = this.clusterState.get();
+        var localNodeId = cs.nodes().getLocalNodeId();
+        var isMasterNode = cs.nodes().isLocalNodeElectedMaster();
+        var routingTable = cs.getRoutingTable();
+        if (routingTable == null) {
+            return List.of();
+        }
+        var result = new ArrayList<ShardInfo>();
+        for (var shardRouting : routingTable.allShards()) {
+            var shardId = shardRouting.shardId();
+            if (localNodeId.equals(shardRouting.currentNodeId()) || (isMasterNode && shardRouting.unassigned())) {
+                var shardStateAndSize = shardStateAndSizeProvider.apply(shardId);
+                if (shardStateAndSize != null) {
+                    result.add(new ShardInfo(shardId.id(),
+                                             shardId.getIndexName(),
+                                             shardRouting.state().name(),
+                                             shardStateAndSize.v1().name(),
+                                             shardStateAndSize.v2()));
+                }
+            }
+        }
+        return result;
+    }
+
+    public static class ShardStateAndSizeProvider implements Function<ShardId, Tuple<IndexShardState, Long>> {
+        private final IndicesService indicesService;
+
+        public ShardStateAndSizeProvider(IndicesService indicesService) {
+            this.indicesService = indicesService;
+        }
+
+        public Tuple<IndexShardState, Long> apply(ShardId shardId) {
+            var shard = indicesService.getShardOrNull(shardId);
+            if (shard == null) {
+                return null;
+            }
+            return new Tuple<>(shard.state(), shard.storeStats().getSizeInBytes());
```

Comment:
`shard.storeStats().getSizeInBytes()` might involve an io operation (if data is not already cached) 
https://github.com/crate/crate/blob/3e7490b0fef082ed3cb8213c7a441129c60fb1ff/server/src/main/java/org/elasticsearch/index/store/ByteSizeCachingDirectory.java#L118
and should might be problematic.